### PR TITLE
Fix #12727 (partial applications in Rec_check)

### LINKED ISSUE #12727: Segmentation fault combining value let rec and labelled partial applications
```text
# let f ~x = let z = x ~y:0 in fun ~y ~u:() -> z + (x ~y);;
val f : x:(y:int -> int) -> y:int -> u:unit -> int = <fun>
# let rec x = f ~x ~u:();;
Segmentation fault
```
`Rec_check` considers that an application with missing labelled parameters is a simple closure allocation, not inspecting the function and arguments. But currently this is wrong, as the function is applied to all parameters until the first missing.

### PR DESCRIPTION
**Pull request by lthls**
This makes `Rec_check` follow the current semantics for labelled applications: all arguments before the first omitted one are passed to the function immediately, all arguments after that are stored in a closure.
There is no particular support for optional arguments, so a few cases end up rejected even though with the current state of the compiler that could have been allowed.
Here is such an example:
```ocaml
let f ?x ~y ~z = ()
let rec x = f ~x ~z:()
```
I don't think this is going to bother anyone.

I suggest @gasche or @yallop as reviewer; this PR is completely independent from my ongoing work around compilation of let-recs, and is a bug in the mode computations (or in how the rest of the compiler handles labelled partial applications, depending on your point of view).

### CODE REVIEW

File: typing/rec_check.ml [hunk 1]
```diff
@@ -577,15 +577,30 @@ let rec expression : Typedtree.expression -> term_judg =
       *)
       expression arg << Guard
     | Texp_apply (e, args)  ->
+        (* [args] may contain omitted arguments, corresponding to labels in
+           the function's type that were not passed in the actual application.
+           The arguments before the first omitted argument are passed to the
+           function immediately, so they are dereferenced. The arguments after
+           the first omitted one are stored in a closure, so guarded.
+           The function itself is called immediately (dereferenced) if there
+           is at least one argument before the first omitted one, stored in
```
Comment (gasche) [inline typing/rec_check.ml]: Nitpick: I find ", stored in the closure otherwise" a bit hard to parse (it is not clear that it is an alternative to "called immediately". I would stop the sentence at the comma and write a new sentence for the second case.

File: typing/rec_check.ml [hunk 2]
```diff
@@ -577,15 +577,30 @@ let rec expression : Typedtree.expression -> term_judg =
...
+           the closure (guarded) otherwise. *)
+        let rec split_args before_omitted = function
+          | (_, None) :: rest -> before_omitted, rest
+          | ((_, Some _) as arg) :: rest ->
+            split_args (arg :: before_omitted) rest
+          | [] -> before_omitted, []
+        in
```
Comment (gasche) [inline typing/rec_check.ml:597]: I am not fond of the way this function is written:

- it is not obvious what `before_omitted` means (it could be "the arguments before the first omitted arguments", which you mean, but it could also be "the omitted arguments that we have seen before the current point"); I propose to fix this by having a non-tail-recursive function without this accumulator parameter
- the `dereferenced` arguments are returned in reverse order; this is innocuous to compute the whole result but it still looks like a potential bug on first reading, I prefer to preserve the order or point out explicitly. I propose to fix this by having a non-tail-recursive function.
- the first omitted argument is silently dropped, when the function name suggest that it keeps all arguments. This does not make a difference in the result but then again it increases the potential for confusion. Please keep it as part of the second list returned.

```ocaml
let rec split_args = function
| [] -> [], []
| ((_, None) :: _) as omitted -> [], omitted
| ((_, Some _) as arg) :: rest ->
    let applied, omitted = split_args rest in
    arg :: applied, omitted
in
let applied, omitted = split_args args in
...
```

File: typing/rec_check.ml [hunk 3]
```diff
@@ -577,15 +577,30 @@ let rec expression : Typedtree.expression -> term_judg =
...
+        let dereferenced, guarded = split_args [] args in
         let arg (_, eo) = option expression eo in
-        let app_mode = if List.exists is_abstracted_arg args
```
Comment (gasche) [inline typing/rec_check.ml:581]: `is_abstracted_args` is local to this file and used in another place that also looks wrong. Please have a look there and maybe also fix it -- the helper function can probably be removed then. (I think that that other place should now check whether the first argument is applied or not.)

**Review by gasche — COMMENTED** [review]
I really liked the explanation comment which I found very clear. I made some change suggestions in comments.

File: typing/rec_check.ml (hunk 3 above)
Comment (lthls) [reply typing/rec_check.ml:581]: The other place (in the classification code) is fine, actually. As long as there is one omitted argument the whole expression will have a closure allocation in tail position.

File: typing/rec_check.ml (hunk 2 above)
Comment (lthls) [reply typing/rec_check.ml:597]: To me, "omitted" means the argument corresponds to a label that was not given. So I want a distinction between the omitted argument, which has no corresponding expression, and the delayed arguments, which can have expressions that will be guarded. But I agree that it feels weird that some omitted args end up in the delayed part but not the first one.
How about that:
```ocaml
let rec split_args ~has_omitted_arg = function
| [] -> [], []
| (_, None) :: rest -> split_args ~has_omitted_arg:true rest
| (_, Some arg) :: rest ->
  let applied, delayed = split_args ~has_omitted_arg rest in
  if has_omitted_arg then applied, arg :: delayed
  else arg :: applied, delayed
in
let applied, delayed = split_args ~has_omitted_arg:false args in
...
```
I could also tweak the name `split_args` to make it clear it drops the ones without an actual expression (`split_actual_args`, for instance).

Comment (gasche) [reply typing/rec_check.ml:597]: Yes, I like your suggestion as well. Having the `delayed` args be expression instead of expression options will also simplify the rest of the code.

Nitpick: I would break a line after `if has_omitted_arg` because the `then` and `else` cases are symmetric so it is nice to keep them aligned.

**Review by gasche — APPROVED** [review]
I approve and will leave the details (minor final code changes + change entry) to @lthls.
